lib/guest_installation_and_configuration_base.pm

```diff
@@ -0,0 +1,1565 @@
+    return $self;
+}
+
+#Create guest instance by assigning values to its parameters but do no install it
+sub create {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->initialize_guest_params;
+    $self->config_guest_params(@_);
+    $self->print_guest_params;
+    return $self;
+}
+
+#Initialize all guest parameters to avoid uninitialized parameters
+sub initialize_guest_params {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->{$_} //= '' foreach (keys %guest_params);
+    $self->{host_ipaddr}      = get_required_var('SUT_IP');
+    $self->{host_name}        = script_output("hostname");
+    $self->{host_domain_name} = script_output("dnsdomainname");
+    $self->{start_run}        = time();
+    return $self;
+}
+
+#Assign real values to guest instance parameters.Reset [guest_name] to guest name used in [guest_metadata] if they are different.Set [guest_version] to get_required_var('VERSION')
+#if it is not given.Set [guest_version_major] and [guest_version_minor] from [guest_version] it they do not match with [guest_version].Set [guest_build] to get_required_var('BUILD')
+#if it is empty and developing [guest_version], or 'GM' if non-developing [guest_version].
+#The subroutine can be called mainly in two different ways:
+#Firstly,config_guest_params can be called in another subroutine for example, create which takes a hash/dictionary as argument.
+#my %testhash = ('key1' => 'value1', 'key2' => 'value2', 'key3' => 'value3'),$self->create(%testhash) which calls $self->config_guest_params(@_).
+#Secondly,config_guest_params can also be called direcly, for example,$self->config_guest_params(%testhash).
+sub config_guest_params {
+    my $self          = shift;
+    my %_guest_params = @_;
+
+    $self->reveal_myself;
+    if ((scalar(@_) % 2 eq 0) and (scalar(@_) gt 0)) {
+        record_info("Configuring guest instance by using following parameters:", Dumper(\%_guest_params));
+        map { $self->{$_} = $_guest_params{$_} } keys %_guest_params;
+    }
+    else {
+        record_info("Can not configure guest instance with empty or odd number of arguments.Mark it as failed.", Dumper(\%_guest_params));
+        $self->record_guest_installation_result('FAILED') if ($self->{guest_installation_result} eq '');
+        return $self;
+    }
+
+    if ((defined $self->{guest_metadata}) and (grep { /name=/ } split(',', $self->{guest_metadata}))) {
+        my $_guest_name_in_metadata = grep { /name=(.*)/ } split(',', $self->{guest_metadata});
+        if (($self->{guest_name} ne $_guest_name_in_metadata) and ($self->{guest_installation_result} eq '')) {
+            record_info("The guest instance has a different name $_guest_name_in_metadata in metadata than $self->{guest_name}.", "Reset guest name to $_guest_name_in_metadata.");
+            $self->{guest_name} = $_guest_name_in_metadata;
+        }
+    }
+
+    if ($self->{guest_version} eq '') {
+        $self->{guest_version} = get_required_var('VERSION');
```

Comment:
'VERSION' is for host version instead of guest. Take it as default is rather risky. In quite some case, we don't install guest and host with same version.

Reply:
The precondition of setting it to host version (Actually it is not host version, it is current developing version. So during 15-SP3, almost all testsuites will have this VERSION being set to 15-SP3) is guest_version is empty. So by default, it is treated as developing guest on developing host during developing SLES release. I think this is reasonable and provides convenience by setting to a default value. If you have other better default version to set, please let me know.

Reply:
In our virt test, it is the HOST version. pls refer to main.pm. for 15sp3 host, 'VERSION' is ok; but for 15sp2 host, 'TARGET_DEVELOPING_VERSION' is the one you would like to use. 

main.pm:
elsif (get_var("VIRT_AUTOTEST")) {
    if (get_var('REPO_0_TO_INSTALL', '')) {
        #Before host installation starts, swtich to version REPO_0_TO_INSTALL if it is set
        #Save VERSION TO TARGET_DEVELOPING_VERSION
        set_var('TARGET_DEVELOPING_VERSION', get_var('VERSION'));
        #Switch to VERSION_TO_INSTALL
        set_var('VERSION', get_var('VERSION_TO_INSTALL'));
    }

Reply:
I know what concerns you. But I already said this is just a way to set it to a default value if nothing given. So better set it to VERSION or current developing version for the most common cases and most popular usage. If you really care about this parameter, you'd better set it currently by yourself. This is just a remedy and user will be well informed by record_info message.

Reply:
I updated this part to including "TARGET_DEVELOPING_VERSION". So emtpy guest_version will be set to developing version anyway.